shared/lib/multichain/chain-agnostic-permission.ts

```diff
@@ -0,0 +1,257 @@
+ * and returns a unique set.
+ *
+ * @param scopesObject - The scopes object to extract accounts from
+ * @returns Array of unique account IDs
+ */
+export function getAllAccountsFromScopesObject(
+  scopesObject: InternalScopesObject,
+): CaipAccountId[] {
+  const allAccounts = new Set<CaipAccountId>();
+
+  Object.values(scopesObject).forEach(({ accounts }) => {
+    accounts.forEach((account) => {
+      allAccounts.add(account);
+    });
+  });
+
+  return Array.from(allAccounts);
+}
+
+export function getAllAccountsFromScopesObjects(
+  scopesObjects: InternalScopesObject[],
+): CaipAccountId[] {
+  return Array.from(
+    new Set([...scopesObjects.flatMap(getAllAccountsFromScopesObject)]),
+  );
+}
+
+/**
+ * Gets all permitted accounts from a CAIP-25 caveat
+ * This extracts all account IDs from both required and optional scopes
+ * and returns a unique set.
+ *
+ * @param caip25CaveatValue - The CAIP-25 caveat value to extract accounts from
+ * @returns Array of unique account IDs
+ */
+export function getAllAccountIdsFromCaip25CaveatValue(
+  caip25CaveatValue: Caip25CaveatValue,
+): CaipAccountId[] {
+  return Array.from(
+    new Set([
+      ...getAllAccountsFromScopesObject(caip25CaveatValue.requiredScopes),
+      ...getAllAccountsFromScopesObject(caip25CaveatValue.optionalScopes),
+    ]),
+  );
+}
+
+/**
+ * Gets all permitted accounts from a CAIP-25 permission
+ * This extracts all account IDs from both required and optional scopes
+ * and returns a unique set.
+ *
+ * @param caip25Permission - The CAIP-25 permission object
+ * @param caip25Permission.caveats - The caveats of the CAIP-25 permission
+ * @returns Array of unique account IDs
+ */
+export function getAllAccountIdsFromPermission(caip25Permission: {
+  caveats: {
+    type: string;
+    value: Caip25CaveatValue;
+  }[];
+}): CaipAccountId[] {
+  const caip25Caveat = getCaip25CaveatFromPermission(caip25Permission);
+  if (!caip25Caveat) {
+    return [];
+  }
+
+  return getAllAccountIdsFromCaip25CaveatValue(caip25Caveat.value);
+}
+
+export function getAllScopesFromScopesObject(
+  scopesObject: InternalScopesObject,
+): InternalScopeString[] {
+  return Object.keys(scopesObject) as InternalScopeString[];
+}
+
+export function getAllScopesFromScopesObjects(
+  scopesObjects: InternalScopesObject[],
+): InternalScopeString[] {
+  return Array.from(
+    new Set([...scopesObjects.flatMap(getAllScopesFromScopesObject)]),
+  );
+}
+
+/**
+ * Gets all scopes (chain IDs) from a CAIP-25 caveat
+ * This extracts all scopes from both required and optional scopes
+ * and returns a unique set.
+ *
+ * @param caip25CaveatValue - The CAIP-25 caveat value to extract scopes from
+ * @returns Array of unique scope strings (chain IDs)
+ */
+export function getAllScopesFromCaip25CaveatValue(
+  caip25CaveatValue: Caip25CaveatValue,
+): CaipChainId[] {
+  const requiredScopes = Object.keys(caip25CaveatValue.requiredScopes);
```

Comment:
Is it valid to put a scope in both required and optional? Since we are trying to deal with duplicates everywhere

Reply:
If not valid, we should throw earlier so we don't need to deal with duplicates everywhere

Reply:
Neither the standard nor our upstream implementation prohibits it: you could describe some methods as required and some as optional...

I tend to think its a level of expressivity that we don't really need and we should disallow but we may need to think through standards implications upstream to resolve that.

I'd be down to just disallow this and throw an error when its attempted @jiexi

Reply:
yeah... i mean until we enforce requiredScopes, we could simply move them all into optionalScopes when we persist them and then add a check in the caveat validator to only allow empty object for requiredScopes?

Yeah, begs the question though, do we need to eat this complexity to support both properties when we don't even enforce one of them

Reply:
Yes, as with a lot of things here I'd prefer to pause and think about it and evaluate our options when we're in a bit less of a rush.

Reply:
Nothing as permanent as a temporary solution 😉